src/esrc.rs

```diff
@@ -0,0 +1,195 @@
+use bls12_381::{pairing, G1Affine, G1Projective, G2Affine, G2Projective, Scalar};
+use rand::{CryptoRng, RngCore};
+use std::ops::Mul;
+
+#[derive(Copy, Clone, Debug)]
+struct EncryptionKey(G1Projective);
+impl Mul<Scalar> for &EncryptionKey {
+    type Output = G1Projective;
+
+    fn mul(self, rhs: Scalar) -> Self::Output {
+        self.0 * rhs
+    }
+}
+impl From<&EncryptionKey> for G1Affine {
+    fn from(enc_key: &EncryptionKey) -> Self {
+        G1Affine::from(enc_key.0)
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+struct DecryptionKey(Scalar);
+impl Mul<&DecryptionKey> for &G1Projective {
+    type Output = G1Projective;
+
+    fn mul(self, rhs: &DecryptionKey) -> Self::Output {
+        self * rhs.0
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+struct CipherText(G1Projective, G1Projective);
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+struct SigningKey(Scalar, Scalar);
+#[derive(Copy, Clone, Debug)]
+struct VerifyKey(G2Projective, G2Projective);
+#[derive(Copy, Clone, Debug)]
+struct Signature(G1Projective, G1Projective, G2Projective, G1Projective);
+
+#[allow(dead_code)]
+fn setup_encryption_keys(rng: &mut (impl RngCore + CryptoRng)) -> (EncryptionKey, DecryptionKey) {
+    let dec_key = Scalar::from(rng.next_u64());
+    let enc_key = G1Affine::generator() * dec_key;
+    (EncryptionKey(enc_key), DecryptionKey(dec_key))
+}
+
+#[allow(dead_code)]
+fn setup_signing_keys(rng: &mut (impl RngCore + CryptoRng)) -> (SigningKey, VerifyKey) {
+    let x0 = Scalar::from(rng.next_u64()).double();
+    let x1 = Scalar::from(rng.next_u64()).double();
+    let x_hat0 = G2Affine::generator() * x0;
+    let x_hat1 = G2Affine::generator() * x1;
+    (SigningKey(x0, x1), VerifyKey(x_hat0, x_hat1))
+}
+
+#[allow(dead_code)]
+fn encrypt(
+    rng: &mut (impl RngCore + CryptoRng),
+    m: G1Affine,
+    enc_key: &EncryptionKey,
+) -> CipherText {
+    let r = Scalar::from(rng.next_u64());
+    let c0 = G1Affine::generator() * r;
+    let c1 = m + (enc_key * r);
+    CipherText(c0, c1)
+}
+
+#[allow(dead_code)]
+fn decrypt(cm: &CipherText, dec_key: &DecryptionKey) -> G1Affine {
+    let CipherText(c0, c1) = cm;
+    let dec = c1 - (c0 * dec_key);
+    G1Affine::from(dec)
+}
+
+#[allow(dead_code)]
+fn randomize(cm: &CipherText, enc_key: &EncryptionKey, r_prime: Scalar) -> CipherText {
+    let CipherText(c0, c1) = cm;
+    let rc0 = c0 + (G1Affine::generator() * r_prime);
+    let rc1 = c1 + (enc_key * r_prime);
+    CipherText(rc0, rc1)
+}
```

Comment:
FWIW, "prime" used with cryptography typically signals a prime number. I know the paper uses `r'`, which I would read as "r prime", but it might help to just call this `r_new`.
```suggestion
#[allow(dead_code)]
fn randomize(cm: &CipherText, enc_key: &EncryptionKey, r_new: Scalar) -> CipherText {
    let CipherText(c0, c1) = cm;
    let rc0 = c0 + (G1Affine::generator() * r_new);
    let rc1 = c1 + (enc_key * r_new);
    CipherText(rc0, rc1)
}
```